domains/pallets/executive/src/lib.rs

```diff
@@ -402,55 +410,128 @@ where
+        <AllPalletsWithSystem as OnFinalize<BlockNumberFor<ExecutiveConfig>>>::on_finalize(
+            block_number,
+        );
     }
 
     /// Wrapped `frame_executive::Executive::apply_extrinsic`.
     ///
     /// Note the storage root in the end.
-    pub fn apply_extrinsic(uxt: Block::Extrinsic) -> ApplyExtrinsicResult {
+    pub fn apply_extrinsic(uxt: ExtrinsicOf<ExecutiveConfig>) -> ApplyExtrinsicResult {
         Pallet::<ExecutiveConfig>::push_root(Self::storage_root());
-        let res = frame_executive::Executive::<
-            System,
-            Block,
-            Context,
-            UnsignedValidator,
-            AllPalletsWithSystem,
-            COnRuntimeUpgrade,
-        >::apply_extrinsic(uxt);
+
+        // apply the extrinsic within another transaction so that changes can be reverted.
+        let res = with_storage_layer(|| {
+            frame_executive::Executive::<
+                ExecutiveConfig,
+                BlockOf<ExecutiveConfig>,
+                Context,
+                UnsignedValidator,
+                AllPalletsWithSystem,
+                COnRuntimeUpgrade,
+            >::apply_extrinsic(uxt.clone())
+            .map_err(|err| DispatchError::Other(err.into()))
+        });
+
+        // apply extrinsic failed with transaction validity error
+        // this could happen for following scenarios
+        // - Bad extrinsic Signature
+        //      This extrinsic will be ignored by the operators during the bundle check
+        //      and marks such bundle as Invalid.
+        // - Extrinsic execution failed
+        //      There are multiple scenarios why this can happen
+        //      - Inherent extrinsic failed. If this happens, we should fail to apply extrinsic
+        //      - Pre and Post dispatch fails. Check the test `test_domain_block_builder_include_ext_with_failed_predispatch`
+        //        why this could happen. If it fail due to this, then we revert the inner storage changes
+        //        but still include extrinsic so that we can clear inconsistency between block body and trace roots.
+        let res = match res {
+            Ok(dispatch_outcome) => Ok(dispatch_outcome),
+            Err(err) => {
+                let encoded = uxt.encode();
+                let (maybe_signer, dispatch_info) =
+                    ExecutiveConfig::ExtrinsicStorageFees::extract_signer(uxt);
+                // if this is not a normal extrinsic, then transaction should not execute
+                // we should fail here.
+                if dispatch_info.class != DispatchClass::Normal {
+                    return Err(TransactionValidityError::Invalid(
+                        InvalidTransaction::BadMandatory,
+                    ));
+                }
+
+                // charge signer for extrinsic storage
+                // TODO: charge user for the storage of extrinsic on consensus chain
```

Comment:
Not sure we should charge the user for pre-dispatch failed extrinsic, this is a different UX compared to other chains.

Reply:
It is also strange that the nonce of the user is unchanged while the balance is reduced silently.

Reply:
> Not sure we should charge the user for pre-dispatch failed extrinsic, this is a different UX compared to other chains.

I disagree, if the extrinsic is stored then user needs to be charged for that storage. Although we do not charge the for the actual execution since there was not execution. This is also another reason why I do not want to use `pallet_transaction_payment` you brought here https://github.com/subspace/subspace/pull/2343/files#r1432863424

> It is also strange that the nonce of the user is unchanged while the balance is reduced silently.

Well, pre dispatch could fail due to the fact that nonce was the same. If there are further transactions in the bundle that uses this next nonce, those transactions would fail for no reason. And this is not failing silently but rather marking the extrinsic as failed with the reason for such failure. If the user were to check the block, it should easier to understand

Reply:
>if the extrinsic is stored then user needs to be charged for that storage.

This a counter-intuitive behavior for users from other chains (even consensus chain) I don't think it is something that can be easily decided by us, and IIUC it is not related to the issue, right?

>Although we do not charge the for the actual execution since there was not execution. This is also another reason why I do not want to use pallet_transaction_payment you brought here

Sorry, I'm not following, `OnChargeTransaction<T>::withdraw_fee(..)` seems much more clean and straightforward than introducing `ExtrinsicStorageFees` to manually `burn` and `note_operator_rewards`

Reply:
> seems much more clean and straightforward than introducing ExtrinsicStorageFees to manually burn and note_operator_rewards

I disagree. Just because it is "cleaner" does not mean it is doing the same thing.
Please look the change where we are taking funds optimistacally and charge whatever is available but Oncharge does not do the same and fail if the exact funds are not available. If you know other func that does the similar behavior, I would be happy to use such function.

Also note Oncharge is also burning the fees underneath through Drop. I would still prefer to take as many funds if not exact available from the user

Reply:
`OnChargeTransaction` may able to do the same with:
```rust
let actual = ExecutiveConfig::Currency::reducible_balance(who, Expendable, force).min(amount);
OnChargeTransaction<T>::withdraw_fee(actual, ..)
```

>Also note Oncharge is also burning the fees underneath through Drop.

I know it is burn under the hook, my point is we should reuse the existing infra as possible.

Reply:
Nope, AFAIK, That is still not possible because the RuntimeCall type is available after the signature and result of such check is opaque to the calle at this point.

I dont see any route to extract such type unless you know of way to extract it without adding additional infra for that

Reply:
>the RuntimeCall type is available after the signature and result of such check is opaque to the calle at this point.

Do you mean we can't use `let xt = uxt.check(&Default::default())?;`?

Reply:
yes that is what I'm saying but you are free to verify that yourself 🤷🏼